### PR TITLE
Issue 19981 add static cache to graphql

### LINKED ISSUE #19981: Add static cache to Graphql
Customers using GraphQL can write really heavy queries- requesting thousands of results that recurse to related content and include asset metadata.  When querying GraphQL, we need a way to cache the results to ensure consistant performance. 
 We need to add a  `cacheTTL` property that the user can provide to cache  the queries results.  When this is specified, the query should be run then stored for the TTL time.  Subsequent identical queries should return the same data from cache, instead of generating it every time.

We need to get with @fmontes and our front end devs to figure out how this should work and how we can make it work easiest with the apollo graphql js client

We should take the params either using the `request.getParameter` or `request.getHeader`

- `dotcachettl` - the cache will have a default ttl, this will override it. We should use a Tuple2 to store the cache response and the expire time. If cacheTTL is <=0, then live results are returned.
- `dotcachekey` optional - if not set, would be the hash of the query and params.  If set, this would force the cache lookup and put to be based on this key

We will have to set up a default cache region in our config - I would say we use the Expiring cache and hold 5000 responses for 10m.

This is an ee only feature.

A prototype of the cache behavior can be found here, though it needs some work to be added to the core code base:

https://github.com/dotcms-plugins/com.dotcms.graphql.cache

### PR DESCRIPTION
**Pull request by dsilvam**
* Added GraphQLCache configured with TimedCacheProvider by default
* Added Interceptor to cache GraphQL responses with the ability to specify different caching options by request parameters or headers
* Added Utility to obtain request parameters and headers case-insensitive
* Added Postman tests for different caching options 
* Fixed a bad catch in GraphQL

### CODE REVIEW

**Review by victoralfaro-dotcms — APPROVED** [review]
LGTM

File: dotCMS/src/main/java/com/dotcms/graphql/GraphQLCache.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+public class GraphQLCache implements Cachable {
+
+
+    private final DotCacheAdministrator cache = CacheLocator.getCacheAdministrator();
+
+    private final Lazy<Boolean> ENABLED_FROM_CONFIG = Lazy.of(()->Config
+            .getBooleanProperty("GRAPHQL_CACHE_RESULT>", true));
+
+    public Optional<String> get(final String key) {
+        if(!canCache()) return Optional.empty();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/graphql/GraphQLCache.java:34]: use { }

Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/graphql/GraphQLCache.java]: done

File: dotCMS/src/main/java/com/dotcms/graphql/GraphqlCacheWebInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+                ? Optional.empty()
+                : graphCache.get(syncKey);
+
+        if (graphResponseFromCache.isPresent()) {
+            writeResponse(response, graphResponseFromCache.get());
+            refreshCacheEntryInBackgroundIfRequested(requestIn, syncKey);
+            return Result.SKIP_NO_CHAIN;
+        }
+
+        synchronized (syncKey) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/graphql/GraphqlCacheWebInterceptor.java:98]: I am not sure if we need this sync, usually cache provider are thread-safe

File: dotCMS/src/main/java/com/dotcms/graphql/GraphqlCacheWebInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        }
+        return true;
+    }
+
+    private void refreshCacheEntryInBackgroundIfRequested(final HttpServletRequest request, final String key) {
+        final Optional<String> backgroundRefreshParam = getParamOrHeader(request,
+                GRAPHQL_CACHE_REFRESH_PARAM);
+
+        if(backgroundRefreshParam.isPresent()
+                && backgroundRefreshParam.get().equalsIgnoreCase("true")) {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/graphql/GraphqlCacheWebInterceptor.java:180]: use the constant in the left side

**Review by jdotcms — APPROVED** [review]

File: dotCMS/src/main/java/com/dotcms/mock/response/MockHttpWriterCaptureResponse.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpServletResponseWrapper;
+import com.dotcms.mock.response.MockResponse;
+
+
+/**
+ * Proxy for HTTPServletResponse
+ *
+ */
+public class MockHttpWriterCaptureResponse extends HttpServletResponseWrapper implements MockResponse {
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/mock/response/MockHttpWriterCaptureResponse.java:15]: I think this doc is not enough

File: dotCMS/src/main/java/com/dotcms/mock/request/HttpRequestReaderWrapper.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import javax.servlet.http.HttpServletRequestWrapper;
+import com.dotmarketing.exception.DotRuntimeException;
+import com.dotmarketing.util.UtilMethods;
+import com.google.common.io.CharStreams;
+
+
+/**
+ * Proxy for HTTPServletRequest
+ *
+ */
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/mock/request/HttpRequestReaderWrapper.java:17]: I think this doc is not enough

**Review by freddyucv — APPROVED** [review]

**Pull request merged**